[GH-9807] Fix: initialize potentially empty collections at the hydration complete

popov-a-e opened the pull request:
Closes https://github.com/doctrine/orm/issues/9807

popov-a-e commented:
Anybody alive here? 🙂

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH9807Test.php:
```diff
@@ -0,0 +1,134 @@
+                $this->_em->getClassMetadata(GH9807Main::class),
+                $this->_em->getClassMetadata(GH9807Join::class),
+            ]
+        );
```
Please use `OrmFunctionalTestCase::createSchemaForModels` instead.

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH9807Test.php:
```diff
@@ -0,0 +1,134 @@
+    private $joins;
+
+    /**
+     * @return Collection
```
```suggestion
     * @return Collection<int, GH9807Join>
```

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH9807Test.php:
```diff
@@ -0,0 +1,134 @@
+     * @Id
+     * @GeneratedValue
+     */
+    protected $id;
```
Why are ids protected in both entities?

greg0ire commented:
Please improve your commit message according to [the contributing guide](https://www.doctrine-project.org/contribute/index.html#working-on-topic-branches).

popov-a-e commented:
Thanks for the comments @greg0ire, I've fixed it

greg0ire commented on lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:
```diff
@@ -44,6 +44,9 @@ class ObjectHydrator extends AbstractHydrator
     /** @var mixed[] */
     private $initializedCollections = [];
 
+    /** @var mixed[] */
```
This is correct but I think you can be more precise.

```suggestion
    /** @var array<string, PersistentCollection> */
```

popov-a-e replied:
I thought about this ofc but didn't want to differ from the same other collections. This file is to refactor anyway, aint' it?

greg0ire replied:
Yes, but I'm likely to be the one doing it. Please don't give me more work :pray:

popov-a-e replied:
Huh, OK, no problem

greg0ire commented on lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:
```diff
@@ -148,6 +151,12 @@ protected function hydrateAllData()
         }
 
+        foreach ($this->uninitializedCollections as $coll) {
+            if (! $coll->isInitialized()) {
```
> the collection will be considered "existed", and "existed" collections are only replaced if REFRESH hint is present

Did you mean to write "initialized"?

popov-a-e replied:
No, there is such a logic that if a Collection is PersistentCollection and initialized it is added to "existing"
```
else {
            // Is already PersistentCollection, and DON'T REFRESH or FETCH-JOIN!
            $this->existingCollections[$oid . $fieldName] = $value;
 ```

greg0ire replied:
I see 👍; please use "existing" then.

greg0ire reviewed (commented):
I forgot to mention that in the contribution guidelines, but can you wrap the body of your commit message (which is great, good job!) at 72 chars?

popov-a-e commented:
> I forgot to mention that in the contribution guidelines, but can you wrap the body of your commit message (which is great, good job!) at 72 chars?

Hm, do you mean leaving only the headline? 72 chars is not enough to unfold the reasoning

greg0ire commented:
> Hm, do you mean leaving only the headline? 

No, I mean wrapping as in adding a linebreak when a line is 72 chars long. With vim properly configured, you can do that with `gqq`

popov-a-e commented:
Got it, thanks for the hint. I misunderstood "wrapping" for "reducing"
I use PHPStorm however. Consider me casual developer for not using vim :D

beberlei commented on lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:
```diff
@@ -148,6 +151,12 @@ protected function hydrateAllData()
             $coll->takeSnapshot();
         }
 
+        foreach ($this->uninitializedCollections as $coll) {
```
I believe you need to reset this in cleanup()

popov-a-e replied:
Indeed, makes sense

SenseException replied:
Shouldn't there be a test for the cleanup of the new property then?

popov-a-e replied:
I believe that the details of the implementation are not something that needs to be tested by itself. There shouldn't be tests that break for on a correctly working class with an unchanged interface. Like there are no tests for other collections or private fields that are getting reset, for example `$existingCollections` or `$initializedCollections`.

popov-a-e replied:
Tests for checking if collections are getting initialized or not in different circumstances are nice to have but those are not related to the original issue.

beberlei reviewed (commented):
Good from my first review

popov-a-e committed:
ObjectHydrator: defer initialization of potentially empty collections

If ObjectHydrator faces an empty row to an uninitialized collection,
it initializes it, to prevent it from querying again (DDC-1526).
However, if that row is the first but not the only in the collection,
the next rows will be ignored, as the collection will be considered
"existing", and "existing" collections are only replaced if REFRESH hint
is present. To prevent it, we defer initialization to the end of the
hydration.

Fixes GH-9807

greg0ire commented:
Thanks @popov-a-e !
